keps/sig-apps/20190226-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,226 @@
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a User of Kubernetes, I should be able to update my StatefulSet, more than one Pod at a time, in a RollingUpdate way, if my Stateful app can tolerate 
+more than one pod being down, thus allowing my update to finish much faster. 
+
+### Implementation Details
+
+#### API Changes
+
+Following changes will be made to the Rolling Update Strategy for StatefulSet.
+
+```go
+// RollingUpdateStatefulSetStrategy is used to communicate parameter for RollingUpdateStatefulSetStrategyType.
+type RollingUpdateStatefulSetStrategy struct {
+	// THIS IS AN EXISTING FIELD
+        // Partition indicates the ordinal at which the StatefulSet should be
+        // partitioned.
+        // Default value is 0.
+        // +optional
+        Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
+
+	// NOTE THIS IS THE NEW FIELD BEING PROPOSED
+	// The maximum number of pods that can be unavailable during the update.
+        // Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+        // Absolute number is calculated from percentage by rounding down.
+        // Defaults to 1.
+        // +optional
+        MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`
+	
+	...
+}
+```
+
+- By Default, if maxUnavailable is not specified, its value will be assumed to be 1 and StatefulSets will follow their old behavior. This
+  will also help while upgrading from a release which doesnt support maxUnavailable to a release which supports this field.
+- If maxUnavailable is specified, it cannot be greater than total number of replicas.
+- If maxUnavailable is specified and partition is also specified, MaxUnavailable cannot be greater than `replicas-partition`
+- podManagementPolicy = OrderedReady
+  If a partition is specified, maxUnavailable will only apply to all the pods which are staged by the partition. Which means all Pods with 
+an ordinal that is greater than or equal to the partition will be updated when the StatefulSet’s .spec.template is updated. Lets say total 
+replicas is 5 and partition is set to 2 and maxUnavailable is set to 2. If the image is changed in this scenario, pods with ordinal 4 and 3 will go
+down at the same time(because of maxUnavailable), once they are running and ready, pods with ordinal 2 will go down. Pods with ordinal 0
+and 1 will remain untouched due the partition.
+- podManagementPolicy = [Parallel](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#parallel-pod-management)
+  here maxUnavailable does not make sense since in this policy, pods are launched or terminated in parallel and do not wait for pods to become
+  running and ready. So we will never know when two terminate the next batch of maxUnavailable pods.
+
+
+#### Implementation
+
+https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L504
```

Comment:
`master` blob will change over time. If you want to include this, you need to pin to a commit.

Reply:
fixed